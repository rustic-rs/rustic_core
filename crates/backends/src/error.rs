use std::{
    num::{ParseIntError, TryFromIntError},
    process::ExitStatus,
    str::Utf8Error,
};

use displaydoc::Display;
use thiserror::Error;

/// [`BackendAccessErrorKind`] describes the errors that can be returned by the various Backends
#[derive(Error, Debug, Display)]
pub enum BackendAccessErrorKind {
    /// backend {0:?} is not supported!
    BackendNotSupported(String),
    /// backend {0} cannot be loaded: {1:?}
    BackendLoadError(String, anyhow::Error),
    /// no suitable id found for {0}
    NoSuitableIdFound(String),
    /// id {0} is not unique
    IdNotUnique(String),
    /// {0:?}
    #[error(transparent)]
    FromIoError(#[from] std::io::Error),
    /// {0:?}
    #[error(transparent)]
    FromTryIntError(#[from] TryFromIntError),
    /// backoff failed: {0:?}
    BackoffError(#[from] backoff::Error<reqwest::Error>),
    /// parsing failed for url: `{0:?}`
    UrlParsingFailed(#[from] url::ParseError),
    /// generic Ignore error: `{0:?}`
    GenericError(#[from] ignore::Error),
    /// creating data in backend failed
    CreatingDataOnBackendFailed,
    /// writing bytes to backend failed
    WritingBytesToBackendFailed,
    /// removing data from backend failed
    RemovingDataFromBackendFailed,
    /// failed to list files on Backend
    ListingFilesOnBackendFailed,
}

/// [`RcloneErrorKind`] describes the errors that can be returned by a backend provider
#[derive(Error, Debug, Display)]
pub enum RcloneErrorKind {
    /// 'rclone version' doesn't give any output
    NoOutputForRcloneVersion,
    /// cannot get stdout of rclone
    NoStdOutForRclone,
    /// rclone exited with `{0:?}`
    RCloneExitWithBadStatus(ExitStatus),
    /// url must start with http:\/\/! url: {0:?}
    UrlNotStartingWithHttp(String),
    /// StdIo Error: `{0:?}`
    #[error(transparent)]
    FromIoError(#[from] std::io::Error),
    /// utf8 error: `{0:?}`
    #[error(transparent)]
    FromUtf8Error(#[from] Utf8Error),
    /// `{0:?}`
    #[error(transparent)]
    FromParseIntError(#[from] ParseIntError),
}

/// [`RestErrorKind`] describes the errors that can be returned while dealing with the REST API
#[derive(Error, Debug, Display)]
pub enum RestErrorKind {
    /// value `{0:?}` not supported for option retry!
    NotSupportedForRetry(String),
    /// parsing failed for url: `{0:?}`
    UrlParsingFailed(#[from] url::ParseError),
    /// requesting resource failed: `{0:?}`
    RequestingResourceFailed(#[from] reqwest::Error),
    /// couldn't parse duration in humantime library: `{0:?}`
    CouldNotParseDuration(#[from] humantime::DurationError),
    /// backoff failed: {0:?}
    BackoffError(#[from] backoff::Error<reqwest::Error>),
    /// Failed to build HTTP client: `{0:?}`
    BuildingClientFailed(reqwest::Error),
    /// joining URL failed on: {0:?}
    JoiningUrlFailed(url::ParseError),
}
