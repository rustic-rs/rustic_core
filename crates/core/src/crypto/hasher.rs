use std::io::{ErrorKind, Read, Result};

use sha2::{Digest, Sha256};

use crate::id::Id;

/// Hashes the given data.
///
/// # Arguments
///
/// * `data` - The data to hash.
///
/// # Returns
///
/// The hash Id of the data.
#[must_use]
pub fn hash(data: &[u8]) -> Id {
    Id::new(Sha256::digest(data).into())
}

/// Hashes the data from a [`Read`]er.
///
/// # Arguments
///
/// * `reader` - The reader to read the data to hash from.
///
/// # Returns
///
/// # Errors
/// - if the reader encounters an error
///
/// The hash Id of the data.
pub fn hash_reader(mut reader: impl Read) -> Result<Id> {
    let mut buffer = [0; 4096];
    let mut hasher = Sha256::default();

    loop {
        match reader.read(&mut buffer) {
            Err(err) => {
                if err.kind() != ErrorKind::Interrupted {
                    break Err(err);
                }
            }
            Ok(count) => {
                if count == 0 {
                    let id = hasher.finalize();
                    break Ok(Id::new(id.into()));
                }
                hasher.update(&buffer[..count]);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use proptest::prelude::*;

    proptest! {
        #[test]
        fn hash_reader_is_identical_to_hash(bytes in prop::collection::vec(prop::num::u8::ANY, 0..65536))  {
            let hash1 = hash(&bytes);
            let hash2 = hash_reader(&*bytes).unwrap();
            prop_assert_eq!(hash1, hash2);
        }
    }
}
