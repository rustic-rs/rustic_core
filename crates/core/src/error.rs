//! Error types and Result module.
//!
//! ## Error handling rules
//!
//! ### Visibility
//!
//! All `pub fn` (associated) functions need to return a `Result<T, RusticError> (==RusticResult<T>)`, if they are fallible.
//! As they are user facing and will cross the API boundary we need to make sure they are high-quality errors containing all
//! needed information and actionable guidance.
//!
//! `pub(crate) fn` visibility should use a local error and thus a Result and error type limited in visibility, e.g.
//! `pub(crate) type ArchiverResult<T> = Result<T, ArchiverErrorKind>`.
//!
//! ### Downgrading and Forwarding
//!
//! `RusticError`s should **not** be downgraded, instead we **upgrade** the function signature to contain a `RusticResult`.
//! For instance, if a function returns `Result<T, ArchiverErrorKind>` and we discover an error path that contains a `RusticError`,
//! we don't need to convert that into an `ArchiverErrorKind`, we should change the function signature, so it returns either a
//! `Result<T, RusticError> (==RusticResult<T>)` or nested results like `RusticResult<Result<T, ArchiverErrorKind>>`.
//! So even if the visibility of that function is `fn` or `pub(crate) fn` it should return a `RusticResult` containing a `RusticError`.
//!
//! If we `map_err` or `and_then` a `RusticError`, we don't want to create a new RusticError from it, but just attach some context
//! to it, e.g. `map_err(|e| e.attach_context("key", "value"))`, so we don't lose the original error. We can also change the error
//! kind with `map_err(|e| e.overwrite_kind(ErrorKind::NewKind))`. If we want to pre- or append to the guidance, we can use
//! `map_err(|e| e.append_guidance_line("new line"))` or `map_err(|e| e.prepend_guidance_line("new line"))`.
//!
//! ### Conversion and Nested Results
//!
//! Converting between different error kinds or their variants e.g. `TreeErrorKind::Channel` -> `ArchiverErrorKind::Channel`
//! should seldom happen (probably never?), as the caller is most likely not setup to handle such errors from a different layer,
//! so at this point, we should return either a `RusticError` indicating this is a hard error. Or use a nested Result, e.g.
//! `Result<Result<T, TreeErrorKind>, RusticError>`.
//!
//! Local error types in `pub fn` (associated) functions need to be manually converted into a `RusticError` with a good error message
//! and other important information, e.g. actionable guidance for the user.
//!
//! ### Backend traits
//!
//! By using `RusticResult` in our `Backend` traits, we also make sure, we get back presentable errors for our users.
//! We had them before as type erased errors, that we just bubbled up. Now we can provide more context and guidance.
//!
//! ### Traits
//!
//! All traits and implementations of (foreign) traits should use `RusticResult` as return type or `Box<RusticError>` as `Self::Err`.
//!
//! ### Display and Debug
//!
//! All types that we want to attach to an error should implement `Display` and `Debug` to provide a good error message and a nice way
//! to display the error.

// FIXME: Remove when 'displaydoc' has fixed/recommended further treatment upstream: https://github.com/yaahc/displaydoc/issues/48
#![allow(clippy::doc_markdown)]
// use std::error::Error as StdError;
// use std::fmt;

use derive_more::derive::Display;
use ecow::{EcoString, EcoVec};
use std::{
    backtrace::{Backtrace, BacktraceStatus},
    convert::Into,
    fmt::{self, Display},
};

pub(crate) mod constants {
    pub const DEFAULT_DOCS_URL: &str = "https://rustic.cli.rs/docs/errors/";
    pub const DEFAULT_ISSUE_URL: &str = "https://github.com/rustic-rs/rustic_core/issues/new";
}

/// Result type that is being returned from methods that can fail and thus have [`RusticError`]s.
pub type RusticResult<T, E = Box<RusticError>> = Result<T, E>;

/// Severity of an error, ranging from informational to fatal.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Display)]
pub enum Severity {
    /// Informational
    Info,

    /// Warning
    Warning,

    /// Error
    Error,

    /// Fatal
    Fatal,
}

/// Status of an error, indicating whether it is permanent, temporary, or persistent.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Display)]
pub enum Status {
    /// Permanent, may not be retried
    Permanent,

    /// Temporary, may be retried
    Temporary,

    /// Persistent, may be retried, but may not succeed
    Persistent,
}

// NOTE:
//
// we use `an error related to {kind}` in the Display impl, so the variant display comments
// should be able to be used in a sentence.
//
/// [`ErrorKind`] describes the errors that can happen while executing a high-level command.
///
/// This is a non-exhaustive enum, so additional variants may be added in future. It is
/// recommended to match against the wildcard `_` instead of listing all possible variants,
/// to avoid problems when new variants are added.
#[non_exhaustive]
#[derive(thiserror::Error, Debug, displaydoc::Display, Default, Clone, Copy, PartialEq, Eq)]
pub enum ErrorKind {
    /// append-only mode
    AppendOnly,
    /// the backend
    Backend,
    /// the configuration
    Configuration,
    /// cryptographic operations
    Cryptography,
    /// running an external command
    ExternalCommand,
    /// internal operations
    // Blob, Pack, Index, Tree Errors
    // Compression, Parsing, Multithreading etc.
    // These are deep errors that are not expected to be handled by the user.
    Internal,
    /// invalid user input
    InvalidInput,
    /// input/output operations
    InputOutput,
    /// a key
    Key,
    /// missing user input
    MissingInput,
    /// general operations
    #[default]
    Other,
    /// password handling
    Password,
    /// the repository
    Repository,
    /// unsupported operations
    Unsupported,
    /// verification
    Verification,
    /// the virtual filesystem
    Vfs,
}

#[derive(thiserror::Error, Debug)]
#[non_exhaustive]
/// Errors that can result from rustic.
pub struct RusticError {
    /// The kind of the error.
    kind: ErrorKind,

    /// The error message with guidance.
    guidance: EcoString,

    /// The URL of the documentation for the error.
    docs_url: Option<EcoString>,

    /// Error code.
    error_code: Option<EcoString>,

    /// Whether to ask the user to report the error.
    ask_report: bool,

    /// The URL of an already existing issue that is related to this error.
    existing_issue_urls: EcoVec<EcoString>,

    /// The URL of the issue tracker for opening a new issue.
    new_issue_url: Option<EcoString>,

    /// The context of the error.
    context: EcoVec<(EcoString, EcoString)>,

    /// Chain to the cause of the error.
    source: Option<Box<(dyn std::error::Error + Send + Sync)>>,

    /// Severity of the error.
    severity: Option<Severity>,

    /// The status of the error.
    status: Option<Status>,

    /// Backtrace of the error.
    ///
    // Need to use option, otherwise thiserror will not be able to derive the Error trait.
    backtrace: Option<Backtrace>,
}

impl Display for RusticError {
    #[allow(clippy::too_many_lines)]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(
            f,
            "`rustic_core` experienced an error related to `{}`.",
            self.kind
        )?;

        writeln!(f)?;
        writeln!(f, "Message:")?;
        let context = if self.context.is_empty() {
            writeln!(f, "{}", self.guidance)?;
            Vec::new()
        } else {
            // If there is context, we want to iterate over it
            // use the key to replace the placeholder in the guidance.
            let mut guidance = self.guidance.to_string();
            let context = self
                .context
                .iter()
                // remove context which has been used in the guidance
                .filter(|(key, value)| {
                    let pattern = "{".to_owned() + key + "}";
                    if guidance.contains(&pattern) {
                        guidance = guidance.replace(&pattern, value);
                        false
                    } else {
                        true
                    }
                })
                .collect();
            writeln!(f, "{guidance}")?;
            context
        };

        if let Some(code) = &self.error_code {
            let default_docs_url = EcoString::from(constants::DEFAULT_DOCS_URL);
            let docs_url = self
                .docs_url
                .as_ref()
                .unwrap_or(&default_docs_url)
                .to_string();

            // If the docs_url doesn't end with a slash, add one.
            let docs_url = if docs_url.ends_with('/') {
                docs_url
            } else {
                docs_url + "/"
            };

            writeln!(f)?;
            writeln!(f, "For more information, see: {docs_url}{code}")?;
        }

        if !self.existing_issue_urls.is_empty() {
            writeln!(f)?;
            writeln!(f, "Related issues:")?;
            self.existing_issue_urls
                .iter()
                .try_for_each(|url| writeln!(f, "- {url}"))?;
        }

        if self.ask_report {
            let default_issue_url = EcoString::from(constants::DEFAULT_ISSUE_URL);
            let new_issue_url = self.new_issue_url.as_ref().unwrap_or(&default_issue_url);

            writeln!(f)?;

            writeln!(
                f,
                "We believe this is a bug, please report it by opening an issue at:"
            )?;
            writeln!(f, "{new_issue_url}")?;
            writeln!(f)?;
            writeln!(
                f,
                "If you can, please attach an anonymized debug log to the issue."
            )?;
            writeln!(f)?;
            writeln!(f, "Thank you for helping us improve rustic!")?;
        }

        writeln!(f)?;
        writeln!(f)?;

        writeln!(f, "Some additional details ...")?;

        if !context.is_empty() {
            writeln!(f)?;
            writeln!(f, "Context:")?;
            context
                .iter()
                .try_for_each(|(key, value)| writeln!(f, "- {key}: {value}"))?;
        }

        if let Some(cause) = &self.source {
            writeln!(f)?;
            writeln!(f, "Caused by:")?;
            writeln!(f, "{cause}")?;
            if let Some(source) = cause.source() {
                write!(f, " : (source: {source:?})")?;
            }
            writeln!(f)?;
        }

        if let Some(severity) = &self.severity {
            writeln!(f)?;
            writeln!(f, "Severity: {severity}")?;
        }

        if let Some(status) = &self.status {
            writeln!(f)?;
            writeln!(f, "Status: {status}")?;
        }

        if let Some(backtrace) = &self.backtrace {
            writeln!(f)?;
            writeln!(f, "Backtrace:")?;
            write!(f, "{backtrace}")?;

            if backtrace.status() == BacktraceStatus::Disabled {
                writeln!(
                    f,
                    " (set 'RUST_BACKTRACE=\"1\"' environment variable to enable)"
                )?;
            }
        }

        Ok(())
    }
}

// Accessors for anything we do want to expose publicly.
impl RusticError {
    /// Creates a new error with the given kind and guidance.
    pub fn new(kind: ErrorKind, guidance: impl Into<EcoString>) -> Box<Self> {
        Box::new(Self {
            kind,
            guidance: guidance.into(),
            context: EcoVec::default(),
            source: None,
            error_code: None,
            docs_url: None,
            new_issue_url: None,
            existing_issue_urls: EcoVec::default(),
            severity: None,
            status: None,
            ask_report: false,
            // `Backtrace::capture()` will check if backtrace has been enabled
            // internally. It's zero cost if backtrace is disabled.
            backtrace: Some(Backtrace::capture()),
        })
    }

    /// Creates a new error with the given kind and guidance.
    pub fn with_source(
        kind: ErrorKind,
        guidance: impl Into<EcoString>,
        source: impl Into<Box<(dyn std::error::Error + Send + Sync)>>,
    ) -> Box<Self> {
        Self::new(kind, guidance).attach_source(source)
    }

    /// Checks if the error has a specific error code.
    pub fn is_code(&self, code: &str) -> bool {
        self.error_code
            .as_ref()
            .map_or(false, |c| c.as_str() == code)
    }

    /// Checks if the error is due to an incorrect password
    pub fn is_incorrect_password(&self) -> bool {
        self.is_code("C002")
    }

    /// Creates a new error from a given error.
    pub fn from<T: std::error::Error + Display + Send + Sync + 'static>(
        kind: ErrorKind,
        error: T,
    ) -> Box<Self> {
        Self::with_source(kind, error.to_string(), error)
    }
}

// Setters for anything we do want to expose publicly.
//
// These were initially generated by `derive_setters`,
// and then manually adjusted to return `Box<Self>` instead of `Self` which
// unfortunately is not possible with the current version of the `derive_setters`.
//
// BEWARE! `attach_context` is manually implemented to allow for multiple contexts
// to be added and is not generated by `derive_setters`.
impl RusticError {
    /// Attach what kind the error is.
    pub fn overwrite_kind(self, value: impl Into<ErrorKind>) -> Box<Self> {
        Box::new(Self {
            kind: value.into(),
            ..self
        })
    }

    /// Ask the user to report the error.
    pub fn ask_report(self) -> Box<Self> {
        Box::new(Self {
            ask_report: true,
            ..self
        })
    }

    /// Attach a chain to the cause of the error.
    pub fn attach_source(
        self,
        value: impl Into<Box<(dyn std::error::Error + Send + Sync)>>,
    ) -> Box<Self> {
        Box::new(Self {
            source: Some(value.into()),
            ..self
        })
    }

    /// Attach the error message with guidance.
    pub fn overwrite_guidance(self, value: impl Into<EcoString>) -> Box<Self> {
        Box::new(Self {
            guidance: value.into(),
            ..self
        })
    }

    /// Append a newline to the guidance message.
    /// This is useful for adding additional information to the guidance.
    pub fn append_guidance_line(self, value: impl Into<EcoString>) -> Box<Self> {
        Box::new(Self {
            guidance: format!("{}\n{}", self.guidance, value.into()).into(),
            ..self
        })
    }

    /// Prepend a newline to the guidance message.
    /// This is useful for adding additional information to the guidance.
    pub fn prepend_guidance_line(self, value: impl Into<EcoString>) -> Box<Self> {
        Box::new(Self {
            guidance: format!("{}\n{}", value.into(), self.guidance).into(),
            ..self
        })
    }

    // IMPORTANT: This is manually implemented to allow for multiple contexts to be added.
    /// Attach context to the error.
    pub fn attach_context(
        mut self,
        key: impl Into<EcoString>,
        value: impl Into<EcoString>,
    ) -> Box<Self> {
        self.context.push((key.into(), value.into()));
        Box::new(self)
    }

    /// Overwrite context of the error.
    ///
    /// # Caution
    ///
    /// This should not be used in most cases, as it will overwrite any existing contexts.
    /// Rather use `attach_context` for multiple contexts.
    pub fn overwrite_context(self, value: impl Into<EcoVec<(EcoString, EcoString)>>) -> Box<Self> {
        Box::new(Self {
            context: value.into(),
            ..self
        })
    }

    /// Attach the URL of the documentation for the error.
    pub fn attach_docs_url(self, value: impl Into<EcoString>) -> Box<Self> {
        Box::new(Self {
            docs_url: Some(value.into()),
            ..self
        })
    }

    /// Attach an error code.
    pub fn attach_error_code(self, value: impl Into<EcoString>) -> Box<Self> {
        Box::new(Self {
            error_code: Some(value.into()),
            ..self
        })
    }

    /// Attach the URL of the issue tracker for opening a new issue.
    pub fn attach_new_issue_url(self, value: impl Into<EcoString>) -> Box<Self> {
        Box::new(Self {
            new_issue_url: Some(value.into()),
            ..self
        })
    }

    /// Attach the URL of an already existing issue that is related to this error.
    pub fn attach_existing_issue_url(mut self, value: impl Into<EcoString>) -> Box<Self> {
        self.existing_issue_urls.push(value.into());
        Box::new(self)
    }

    /// Attach the severity of the error.
    pub fn attach_severity(self, value: impl Into<Severity>) -> Box<Self> {
        Box::new(Self {
            severity: Some(value.into()),
            ..self
        })
    }

    /// Attach the status of the error.
    pub fn attach_status(self, value: impl Into<Status>) -> Box<Self> {
        Box::new(Self {
            status: Some(value.into()),
            ..self
        })
    }

    /// Overwrite the backtrace of the error.
    ///
    /// This should not be used in most cases, as the backtrace is automatically captured.
    pub fn overwrite_backtrace(self, value: impl Into<Backtrace>) -> Box<Self> {
        Box::new(Self {
            backtrace: Some(value.into()),
            ..self
        })
    }
}
