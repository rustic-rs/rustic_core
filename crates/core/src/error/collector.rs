use derive_more::From;

use crate::{error::RusticErrorKind, RusticError};
use log::{error, info, warn};

#[derive(Debug)]
pub enum RusticIssue {
    /// An error issue, indicating that something went wrong irrecoverably
    Error(RusticError),

    /// A warning issue, indicating that something might be wrong
    Warning(RusticWarning),

    /// An info issue, indicating additional information
    Info(RusticInfo),
}

impl RusticIssue {
    pub fn new_error(error: RusticErrorKind) -> Self {
        Self::Error(error.into())
    }

    pub fn new_warning(message: &str) -> Self {
        Self::Warning(message.into())
    }

    pub fn new_info(message: &str) -> Self {
        Self::Info(message.into())
    }

    pub fn log(&self) {
        match self {
            Self::Error(error) => error!("{}", error),
            Self::Warning(warning) => warn!("{}", warning.0),
            Self::Info(info) => info!("{}", info.0),
        }
    }
}

/// A rustic warning message
///
/// Warning messages are used to indicate that something might be wrong.
#[derive(Debug, Clone, From)]
pub struct RusticWarning(String);

impl RusticWarning {
    pub fn new(message: &str) -> Self {
        Self(message.to_owned())
    }
}

impl From<&str> for RusticWarning {
    fn from(message: &str) -> Self {
        Self::new(message)
    }
}

/// A rustic info message
///
/// Info messages are used to provide additional information to the user.
#[derive(Debug, Clone, From)]
pub struct RusticInfo(String);

impl RusticInfo {
    pub fn new(message: &str) -> Self {
        Self(message.to_owned())
    }
}

impl From<&str> for RusticInfo {
    fn from(message: &str) -> Self {
        Self::new(message)
    }
}

#[derive(Debug, Default)]
pub struct RusticIssueCollector {
    /// The errors collected
    errors: Option<Vec<RusticError>>,

    /// The warnings collected
    warnings: Option<Vec<RusticWarning>>,

    /// The info collected
    info: Option<Vec<RusticInfo>>,

    /// Whether to log items directly during addition
    log: bool,
}

impl RusticIssueCollector {
    pub fn new(log: bool) -> Self {
        Self {
            errors: None,
            warnings: None,
            info: None,
            log,
        }
    }

    pub fn add(&mut self, issue: RusticIssue) {
        match issue {
            RusticIssue::Error(error) => self.add_error(error.0),
            RusticIssue::Warning(warning) => self.add_warning(&warning.0),
            RusticIssue::Info(info) => self.add_info(&info.0),
        }
    }

    pub fn add_error(&mut self, error: RusticErrorKind) {
        if self.log {
            error!("{error}");
        }

        if let Some(errors) = &mut self.errors {
            errors.push(error.into());
        } else {
            self.errors = Some(vec![error.into()]);
        }
    }

    pub fn add_warning(&mut self, message: &str) {
        if self.log {
            warn!("{message}");
        }

        if let Some(warnings) = &mut self.warnings {
            warnings.push(message.to_owned().into());
        } else {
            self.warnings = Some(vec![message.to_owned().into()]);
        }
    }

    pub fn add_info(&mut self, message: &str) {
        if self.log {
            warn!("{message}");
        }

        if let Some(info) = &mut self.info {
            info.push(message.to_owned().into());
        } else {
            self.info = Some(vec![message.to_owned().into()]);
        }
    }

    pub fn has_errors(&self) -> bool {
        self.errors.is_some()
    }

    pub fn has_warnings(&self) -> bool {
        self.warnings.is_some()
    }

    pub fn has_info(&self) -> bool {
        self.info.is_some()
    }

    pub fn get_errors(&self) -> Option<Vec<&RusticError>> {
        self.errors.as_ref().map(|errors| errors.iter().collect())
    }

    pub fn get_warnings(&self) -> Option<Vec<RusticWarning>> {
        self.warnings.clone()
    }

    pub fn get_info(&self) -> Option<Vec<RusticInfo>> {
        self.info.clone()
    }

    pub fn log_all(&self) {
        self.log_all_errors();
        self.log_all_warnings();
        self.log_all_info();
    }

    pub fn log_all_errors(&self) {
        if let Some(errors) = &self.errors {
            for error in errors {
                error!("{}", error);
            }
        }
    }

    pub fn log_all_warnings(&self) {
        if let Some(warnings) = &self.warnings {
            for warning in warnings {
                warn!("{}", warning.0);
            }
        }
    }

    pub fn log_all_info(&self) {
        if let Some(info) = &self.info {
            for info in info {
                info!("{}", info.0);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::error::RusticErrorKind;

    #[test]
    fn test_add_issue() {
        let mut collector = RusticIssueCollector::default();

        let issue = RusticIssue::new_error(RusticErrorKind::StdIo(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            "test",
        )));

        collector.add(issue);
        assert!(collector.has_errors());
        assert!(!collector.has_warnings());
        assert!(!collector.has_info());
    }

    #[test]
    fn test_add_error() {
        let mut collector = RusticIssueCollector::default();
        collector.add_error(RusticErrorKind::StdIo(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            "test",
        )));

        assert!(collector.has_errors());
        assert!(!collector.has_warnings());
        assert!(!collector.has_info());
    }

    #[test]
    fn test_add_warning() {
        let mut collector = RusticIssueCollector::default();
        collector.add_warning("test");
        assert!(!collector.has_errors());
        assert!(collector.has_warnings());
        assert!(!collector.has_info());
    }

    #[test]
    fn test_add_info() {
        let mut collector = RusticIssueCollector::default();
        collector.add_info("test");
        assert!(!collector.has_errors());
        assert!(!collector.has_warnings());
        assert!(collector.has_info());
    }

    #[test]
    fn test_get_errors() {
        let mut collector = RusticIssueCollector::default();
        collector.add_error(RusticErrorKind::StdIo(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            "test",
        )));
        assert_eq!(collector.get_errors().unwrap().len(), 1);
    }

    #[test]
    fn test_get_warnings() {
        let mut collector = RusticIssueCollector::default();
        collector.add_warning("test");
        assert_eq!(collector.get_warnings().unwrap().len(), 1);
    }

    #[test]
    fn test_get_info() {
        let mut collector = RusticIssueCollector::default();
        collector.add_info("test");
        assert_eq!(collector.get_info().unwrap().len(), 1);
    }
}
