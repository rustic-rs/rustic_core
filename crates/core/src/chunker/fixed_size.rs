use std::io::Read;

use crate::error::{ErrorKind, RusticError, RusticResult};

/// `ChunkIter` is an iterator that chunks data.
pub(crate) struct ChunkIter<R: Read + Send> {
    /// The reader.
    reader: R,

    /// The size hint is used to optimize memory allocation; this should be an upper bound on the size.
    size_hint: usize,

    /// The size of a chunk.
    size: usize,

    /// If the iterator is finished.
    finished: bool,
}

impl<R: Read + Send> ChunkIter<R> {
    /// Creates a new `ChunkIter`.
    ///
    /// # Arguments
    ///
    /// * `reader` - The reader to read from.
    /// * `size_hint` - The size hint is used to optimize memory allocation; this should be an upper bound on the size.
    /// * `rabin` - The rolling hash.
    pub(crate) fn new(size: usize, reader: R, size_hint: usize) -> Self {
        Self {
            reader,
            size_hint, // size hint is used to optimize memory allocation; this should be an upper bound on the size
            size,
            finished: false,
        }
    }
}

impl<R: Read + Send> Iterator for ChunkIter<R> {
    type Item = RusticResult<Vec<u8>>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.finished {
            return None;
        }

        let mut vec = Vec::with_capacity(self.size_hint.min(self.size));

        let size = match (&mut self.reader)
            .take(self.size as u64)
            .read_to_end(&mut vec)
        {
            Ok(size) => size,
            Err(err) => {
                return Some(Err(RusticError::with_source(
                    ErrorKind::InputOutput,
                    "Failed to read from reader in iterator",
                    err,
                )));
            }
        };

        // If self.min_size is not reached, we are done.
        // Note that the read data is of size size + open_buf_len and self.min_size = minsize + open_buf_len
        if size < self.size {
            self.finished = true;
            vec.truncate(size);
        }
        self.size_hint = self.size_hint.saturating_sub(vec.len()); // size_hint can be too small!
        if vec.is_empty() { None } else { Some(Ok(vec)) }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::crypto::hasher::hash;
    use insta::assert_ron_snapshot;
    use rand::prelude::*;
    use rstest::rstest;
    use std::io::Cursor;

    #[rstest]
    #[case(1024*1024)]
    #[case(1021*1024)]
    fn chunk_random(#[case] chunk_size: usize) {
        const RANDOM_DATA_SIZE: usize = 32 * 1024 * 1024;
        const SEED: u64 = 23;
        let mut rng = StdRng::seed_from_u64(SEED);
        let mut data = vec![0u8; RANDOM_DATA_SIZE];
        rng.fill_bytes(&mut data);
        let mut reader = Cursor::new(data);

        let chunker = ChunkIter::new(chunk_size, &mut reader, 0);
        let chunks: Vec<_> = chunker
            .map(|chunk| {
                let chunk = chunk.unwrap();
                (chunk.len(), hash(&chunk))
            })
            .collect();

        assert_ron_snapshot!(format!("chunk-size{chunk_size}"), chunks);
    }
}
