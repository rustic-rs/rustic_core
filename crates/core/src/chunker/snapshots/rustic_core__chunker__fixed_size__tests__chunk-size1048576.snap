---
source: crates/core/src/chunker/fixed_size.rs
expression: chunks
---
[
  (1048576, Id("edde0629189f5682f9f0c320318e536a30891cd55e12a2507cc11fa0c9aae0a2")),
  (1048576, Id("55430b3d2960f60336ea9b42cd00e31497fa20d797aee9610b2a8df0b8903537")),
  (1048576, Id("95723a06d809591d21fdd4926e8eb9c4c5ce66031595f27776e73d96c582c9a3")),
  (1048576, Id("6a116f1fe2d131fcf329846881d1505425b39b4ec3fee012f990eaf161bc4d63")),
  (1048576, Id("d0b9ce455df46f7bcd0561e9e6a97e82fa753508b440a86e8c6262023875041f")),
  (1048576, Id("abf6108e84951cbbe4ccb25247af536622d29599ed6cb84eff454da1fc5aee00")),
  (1048576, Id("74f0c45d568ca47275ca9b01ce7b5657d6ef53698cce0836e401012bad157315")),
  (1048576, Id("62685b5db94229efe7aa533025dad9cc5a0721ac6453b863d67ddd488f3e36aa")),
  (1048576, Id("e7bcb1d9452f29b54ea75da82824711fc44ee162886d34329d4ed00f40f4d2a9")),
  (1048576, Id("059730989bc9b1a0e30a1f055c4503794933f3d2086dc215e41dbe4e9f135be5")),
  (1048576, Id("7e2f20f4f43ee8ca44a610a0ac489e2ecbb7d5c4369f87115060a1d1bdf9a160")),
  (1048576, Id("2574b7b56ddda8b3dfa1a53d971bf182ed62d62f4697d853ba94735871b06168")),
  (1048576, Id("11de38819350185c6c3c001393da5fa4dc2dcd5cc8a3d48cd993995df0e43397")),
  (1048576, Id("c0223cf4afde5346d64bf3d1e954fed24a57db161359a7a35ac1ad2cb0b1b141")),
  (1048576, Id("71543b40052ebc6ae7c63e559799c248f78fef71b515fb0820456ffb167b9132")),
  (1048576, Id("8cb18dbdbdada9ddc5180266e91a3a289c1c68dc8bf19bc6dd321ecea14b3f26")),
  (1048576, Id("a2c004688b026c70e30763eb683f2643eaa79ab528f56434be5b61dc8f132887")),
  (1048576, Id("acfe8a27d47c051622e3f4e47fdd4c6f9d11a02462f9b5b9301e68276b8b1614")),
  (1048576, Id("71936eab0103916f7f9ed7ff0c6e2253dcd4f7e0b59d126e69cbefb295c9cd25")),
  (1048576, Id("ff9fc8ffecf7c497da30f2300369c5739f47a15869d261d3107c2dfa34d73c76")),
  (1048576, Id("7fa47bb1731bc1f35e170cf18217329fa7c96b33ffcad4671e1b546f53783f4c")),
  (1048576, Id("82e2477f0e75129f74dcafcfa3b51e6c3b86e5c8790af52c090c75e32e0dd69f")),
  (1048576, Id("26947381e67abb5da2ccb691eac4b8b1511a0b9d03d6d744831bb8426b686487")),
  (1048576, Id("6466c01ae8c15bb62cc90c2419681ffd3e032b6866915d03f38f852ac777f4a2")),
  (1048576, Id("57e715852646ef414ee91eeb0c9d805c81e7cc50aae426bc125f5eba8b6e0c27")),
  (1048576, Id("85e03f0040567a4749960f7c50301a2e07150cdcfc6b01095325fe31764bed2b")),
  (1048576, Id("7f98b9889cd375f120cb70e96ba2d78ac94c350e94a61284e0c23213731b14f3")),
  (1048576, Id("211682df2c183f9b9a6a7fb3c14547b84402632822fb48e737fc7dcb783882fe")),
  (1048576, Id("7e35e2222be5d721ebb0fe1a648718b22b4e25ae35d1e85823f7aac68118c5ba")),
  (1048576, Id("eacc1d9b5fca60d1201c9ddcefadde6cf962c74e6d85c00d79485b1dc825a964")),
  (1048576, Id("2878fa450f0e5b4a3fe952b66183213d85975b7ecad8b2a14c8cc26554423f0e")),
  (1048576, Id("ad9ed85a037d48f10b665d9f47e669e0c5a060c06b24782d5e44e3324e37be37")),
]
